designs/0005-static-matrices.md

```diff
@@ -0,0 +1,444 @@
+
+The idea is that when part of a `varmat` is assigned to, the values in that part of the
+`varmat` are saved into the memory arena before being overwritten, and a call is pushed
+onto the chaining stack that will restore the values back from the arena and set the
+respective adjoints of the `varmat` entries to zero.
+
+The assumptions that makes this work is after a subset of a `varmat` is overwritten,
+that part of the `varmat` cannot be used in any further calculations, and so the adjoints
+of that part of the matrix will be zero in the reverse pass at the time of assignment.
+
+This means that assignment into a `varmat` variable would make it so that views
+point into that part of the `varmat` variable's memory no longer point to the expected
+variables.
+
+# Stanc3 implementation
+[stanc3-implementation]: #stanc3-implementation
+
+The stanc3 implementation goal is to implement `varmat` types in a way that
+is invisible to the user. This means the compiler decides how Stan vectors, row
+vectors, and matrices (or arrays of these types) are implemented automatically,
+either picking an appropriate `varmat` or a `matvar` type, and these decisions should
+be made in a way to avoid making programs slower than they would have been
+with just `matvar` types. For brevity, any Stan vector, row vector, or matrix
+type will be referred to from hereon as just a matrix type. Consider a linear
+regression model:
+
+```stan
+data {
+  int N;
+  int K;
+  matrix[N, K] X;
+  vector[N] y;
+}
+
+parameters {
+  vector[K] b;
+  real<lower = 0.0> sigma;
+}
+
+model {
+  vector[N] mu = X * b;
+  target += normal_lpdf(y | mu, sigma);
+}
+```
+
+In this program there are four named matrix variables, `X`, `y`, `b`, and `mu`.
+
+It is immediately evident that because there are no autodiff types in the
+`data` block, all the matrix types in there are implemented with doubles in
+Eigen, so there is no `matvar` or `varmat` to worry about. The same thing
+applies for `transformed data` and the `generated quantities` block.
+
+This leaves `b` and `mu`, matrix variables defined in the `parameters` and
+`model` block. Any matrix variable defined in a `transformed parameters` block
+would be decided the same way.
+
+The algorithm for figuring out the underlying matrix types for these variables
+is as follows:
+
+1. Define all user-defined functions only for `matvar` types, and all
+internally defined variables and returned variables are `matvar` types as well.
+
+2. Assume every named and unnamed matrix variable in the `parameters`,
+`transformed parameters` and `model` blocks use `varmat` types
+
+3. If a matrix variable is used in a function that does not support `varmat`
+arguments, it is made a `matvar` variable.
+
+4. If single elements of a matrix variable are indexed, it is made a `matvar`
+variable.
+
+5. If a `matvar` matrix variable is assigned to a second matrix variable,
+the second is made to be a `matvar` variable.
```

Comment:
What about assigning a `varmat` to a `matvar`? Isn't that quite expensive too?

Reply:
The way its envisioned now, that can not happen. The variable being assigned to will be labeled matvar and will be matvar for its entire "lifetime". Its a bit conservative.

Reply:
I'm thinking of something like
```stan
parameters {
  matrix[N,K] x;
  matrix[K,L] y;
}
transformed parameters {
  matrix[N,L] z = x * y;
  print(z[1,1]);
}
```
`z` is `matvar` due to indexing. But what about `x` and `y`? Maybe they're `varmat` and the product is calculated as a `varmat` but then converted to a `matvar`. Or are _both_ `x` and `y` forced to be `matvar`? Or only one?

Reply:
Current plan is x and y are forced to be matvar. Mostly because it makes the rules simple. We tried to avoid stuff like: "if X has more than 5 single element access then matvar". We realize these cases are a bit unfortunate.

We circled around this one a bit, so if you have any suggestions, we are all ears.

Reply:
Yeah, I don't have any ideas. I'm just trying to clarify the inference rules in this section.
It says earlier that if a function takes _any_ `varmat` argument then the return type is `varmat`. So conversely, if the return type needs to be `matvar` then all inputs must also be `matvar`.

Reply:
That is definitely a good point yes. That should be clarified.

Reply:
Yeah I think we need another rule for this.

Reply:
Added a rule